### PR TITLE
Update proto and add custom build steps

### PR DESCRIPTION
**Pull request by maiqbal11**
Needed to unblock https://github.com/Azure/azure-functions-python-worker/pull/554. Contains the same logic as https://github.com/Azure/azure-functions-python-worker/pull/528 which was reverted.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -73,56 +75,74 @@ def _gen_grpc(self):
...
-        built_proto_files_dir = (built_protos_dir
-                                 / 'azure_functions_worker' / 'protos')
+        built_protos_dir = build_dir / 'built_protos'
 
-        if os.path.exists(build_dir):
+        if os.path.exists(staging_root_dir):
             shutil.rmtree(build_dir)
 
         if os.path.exists(built_protos_dir):
             shutil.rmtree(built_protos_dir)
```
Comment (Hazhzeng) [inline setup.py]: Seems like these four lines can be combined together?
```python
if os.path.exists(build_dir):
    shutil.rmtree(build_dir)
```
Am I missing something here?

File: setup.py [hunk 2]
```diff
@@ -73,56 +75,74 @@ def _gen_grpc(self):
...
                   file=sys.stderr)
             sys.exit(1)
 
-        for f in compiled:
-            shutil.copy(f, proto_root_dir)
+        # Needed to support absolute imports in files. See
+        # https://github.com/protocolbuffers/protobuf/issues/1491
+        self.make_absolute_imports(compiled_files)
+
+        dir_util.copy_tree(built_protos_dir, str(proto_root_dir))
```
Comment (Hazhzeng) [inline setup.py]: Do we still need **built_protos_dir**? I think we should clean it up after we build it.

File: setup.py [hunk 3]
```diff
@@ -73,56 +75,74 @@ def _gen_grpc(self):
...
-        if os.path.exists(build_dir):
+        if os.path.exists(staging_root_dir):
             shutil.rmtree(build_dir)
 
         if os.path.exists(built_protos_dir):
             shutil.rmtree(built_protos_dir)
 
-        proto_files = glob.glob(str(proto_src_dir / '**' / '*.proto'),
-                                recursive=True)
+        shutil.copytree(proto_src_dir, staging_dir)
```
Comment (Hazhzeng) [inline setup.py]: Doesn't seems like **staging_dir** is used anywhere else. Maybe we can remove it?

Comment (maiqbal11) [reply setup.py]: It's being used to do the copying as you pointed here. In the actual build command, we are using the same directory structure without explicitly referencing the `staging_dir`. Might be useful to walk through this offline as it's slightly tedious to explain here.

**Pull request merged**